Cubical/Structures/Poset.agda

```diff
@@ -0,0 +1,189 @@
+
+open import Cubical.Foundations.Prelude
+open import Cubical.Foundations.Logic
+open import Cubical.Foundations.Isomorphism
+open import Cubical.Foundations.Equiv        hiding   (_■)
+open import Cubical.Foundations.SIP          renaming (SNS-≡ to SNS)
+open import Cubical.Core.Primitives
+open import Cubical.Foundations.HLevels
+open import Cubical.Data.Sigma
+open import Cubical.Data.Sigma.Properties
+
+-- We will adopt the convention of denoting the level of the carrier
+-- set by ℓ₀ and the level of the relation result by ℓ₁.
+private
+  variable
+    ℓ₀ ℓ₁ : Level
+
+Order : (ℓ₁ : Level) → Type ℓ₀ → Type (ℓ-max ℓ₀ (ℓ-suc ℓ₁))
+Order ℓ₁ A = A → A → hProp ℓ₁
+
+order-iso : (M N : Σ (Type ℓ₀) (Order ℓ₁)) →  fst M ≃ fst N → Type (ℓ-max ℓ₀ ℓ₁)
+order-iso (A , _⊑₀_) (B , _⊑₁_) eqv =
+  (x y : A) → [ x ⊑₀ y ⇔ f x ⊑₁ f y ]
+  where
+    f = equivFun eqv
+
+isSetOrder : (ℓ₁ : Level) (A : Type ℓ₀) → isSet (Order ℓ₁ A)
+isSetOrder ℓ₁ A = isSetΠ (λ _ → isSetΠ λ _ → isSetHProp)
+
+Order-is-SNS : SNS {ℓ₀} (Order ℓ₁) order-iso
+Order-is-SNS {ℓ₁ = ℓ₁} {X = X}  _⊑₀_ _⊑₁_ = f , f-equiv
+  where
+    f : order-iso (X , _⊑₀_) (X , _⊑₁_) (idEquiv X) → _⊑₀_ ≡ _⊑₁_
+    f i = funExt λ x → funExt λ y → ⇔toPath (fst (i x y)) (snd (i x y))
+
+    ⇔-prop : isProp ((x y : X) → [ x ⊑₀ y ⇔ x ⊑₁ y ])
+    ⇔-prop = isPropΠ λ x → isPropΠ λ y → snd (x ⊑₀ y ⇔ x ⊑₁ y)
+
+    f-equiv : isEquiv f
+    f-equiv = record { equiv-proof = λ eq → (g eq , sec eq) , h eq }
+      where
+        g : (eq : _⊑₀_ ≡ _⊑₁_)
+          → (x y : X)
+          → ([ x ⊑₀ y ] → [ x ⊑₁ y ]) × ([ x ⊑₁ y ] → [ x ⊑₀ y ])
+        g eq x y = subst (λ { _⊑⋆_ → [ x ⊑⋆ y ] }) eq
+                 , subst (λ { _⊑⋆_ → [ x ⊑⋆ y ] }) (sym eq)
+
+        sec : section f g
+        sec p = isSetOrder _ X _⊑₀_ _⊑₁_ (f (g p)) p
+
+        h : (p : _⊑₀_ ≡ _⊑₁_) → (fib : fiber f p) → (g p , sec p) ≡ fib
+        h p (i , _) = ΣProp≡ A-prop B-prop
+          where
+            A-prop : _
+            A-prop = λ i′ → isOfHLevelSuc 2 (isSetOrder ℓ₁ X) _⊑₀_ _⊑₁_ (f i′) p
+
+            B-prop : _
+            B-prop = ⇔-prop (g p) i
+
+-- We now write down the axioms for a partial order.
+
+isReflexive : {A : Type ℓ₀} → Order ℓ₁ A → hProp (ℓ-max ℓ₀ ℓ₁)
+isReflexive {A = X} _⊑_ =
+  ((x : X) → [ x ⊑ x ]) , isPropΠ (λ x → snd (x ⊑ x))
+
+isTransitive : {A : Type ℓ₀} → Order ℓ₁ A → hProp (ℓ-max ℓ₀ ℓ₁)
+isTransitive {ℓ₀ = ℓ₀} {ℓ₁ = ℓ₁} {A = X} _⊑_ = φ , φ-prop
+  where
+    φ      : Type (ℓ-max ℓ₀ ℓ₁)
+    φ      = ((x y z : X) → [ x ⊑ y ⇒ y ⊑ z ⇒ x ⊑ z ])
+    φ-prop : isProp φ
+    φ-prop = isPropΠ λ x → isPropΠ λ y → isPropΠ λ z →
+               snd (x ⊑ y ⇒ y ⊑ z ⇒ x ⊑ z)
+
+isAntisym : {A : Type ℓ₀} → isSet A → Order ℓ₁ A → hProp (ℓ-max ℓ₀ ℓ₁)
+isAntisym {ℓ₀ = ℓ₀} {ℓ₁ = ℓ₁} {A = X} A-set _⊑_ = φ , φ-prop
+  where
+    φ      : Type (ℓ-max ℓ₀ ℓ₁)
+    φ      = ((x y : X) → [ x ⊑ y ] → [ y ⊑ x ] → x ≡ y)
+    φ-prop : isProp φ
+    φ-prop = isPropΠ λ x → isPropΠ λ y → isPropΠ λ _ → isPropΠ λ _ → A-set x y
```

Comment:
I think we need `isPropΠ4`...

Reply:
Do you want me to add that? I think three is a good number to stop...

Reply:
Yeah, probably. I guess you could still use `isPropΠ3` though